universal-application-tool-0.0.1/app/repository/ApplicationRepository.java

```diff
@@ -5,62 +5,169 @@
 
 import io.ebean.Ebean;
 import io.ebean.EbeanServer;
-import java.time.Clock;
+import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
 import javax.inject.Inject;
 import models.Applicant;
 import models.Application;
+import models.LifecycleStage;
 import models.Program;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import play.db.ebean.EbeanConfig;
+import services.applicant.ApplicantNotFoundException;
+import services.program.ProgramNotFoundException;
 
 public class ApplicationRepository {
   private final ProgramRepository programRepository;
   private final ApplicantRepository applicantRepository;
-  private final Clock clock;
   private final EbeanServer ebeanServer;
   private final DatabaseExecutionContext executionContext;
+  private static final Logger LOG = LoggerFactory.getLogger(ApplicationRepository.class);
 
   @Inject
   public ApplicationRepository(
       ProgramRepository programRepository,
       ApplicantRepository applicantRepository,
-      Clock clock,
       EbeanConfig ebeanConfig,
       DatabaseExecutionContext executionContext) {
-    this.programRepository = programRepository;
-    this.applicantRepository = applicantRepository;
-    this.clock = clock;
+    this.programRepository = checkNotNull(programRepository);
+    this.applicantRepository = checkNotNull(applicantRepository);
     this.ebeanServer = Ebean.getServer(checkNotNull(ebeanConfig).defaultServer());
     this.executionContext = checkNotNull(executionContext);
   }
 
+  /**
+   * Submit an application, which will delete any in-progress drafts, obsolete any submitted
+   * applications to a program with the same name (to include past versions of the same program),
+   * and create a new application in the active state.
+   */
   public CompletionStage<Application> submitApplication(Applicant applicant, Program program) {
     return supplyAsync(
         () -> {
-          Application application = new Application(applicant, program, clock.instant());
-          application.save();
-          return application;
-        });
+          return submitApplicationInternal(applicant, program);
+        },
+        executionContext.current());
   }
 
   public CompletionStage<Optional<Application>> submitApplication(
       long applicantId, long programId) {
+    return this.perform(
+        applicantId,
+        programId,
+        (ApplicationArguments appArgs) ->
+            submitApplicationInternal(appArgs.applicant, appArgs.program));
+  }
+
+  private Application submitApplicationInternal(Applicant applicant, Program program) {
+    ebeanServer.beginTransaction();
+    try {
+      List<Application> oldApplications =
+          ebeanServer
+              .createQuery(Application.class)
+              .where()
+              .eq("applicant.id", applicant.id)
+              .eq("program.name", program.getProgramDefinition().name())
```

Comment:
issue (blocking): afaik we don't do anything to ensure that `programs` with the same name are actually the same program, it's just a text field the admin specifies. What assumptions are being made about programs that share a name here?

Reply:
That they are different versions of the same program - I'll lock that field in / grey it out when clicking a "new version" button, and disallow v1s of programs with the same name as existing ones.

Reply:
Will this prevent programs from being renamed?

Reply:
Unless we do some extra work, yeah, a program with a new name restarts its identity at v1.  We could track them differently, I suppose, do a "previous version" id pointer, maybe, if that's not acceptable.

Reply:
chatted about this, resolved